### PR TITLE
Combine "can this be" and "make this a" boot disk logic

### PR DESCRIPTION
**Pull request by mwhudson**
This is all a bit crazy but it think it will make the conversion of this logic to the v2 partitioning world a lot more self-contained.

### CODE REVIEW

File: subiquity/common/filesystem/tests/test_manipulator.py [hunk 1]
```diff
@@ -234,3 +236,201 @@ def test_add_boot_has_valid_offset(self):
...
+        self.assertIsBootDisk(manipulator, disk)
+        [p1, p2] = sorted(disk.partitions(), key=lambda p: p.offset)
+        size_after = p1.size
+        self.assertIs(p1, part)
+        self.assertEqual(p1.offset, MiB)
+        self.assertEqual(p2.offset, p1.offset + p1.size)
+        self.assertEqual(p2.flag, "prep")
+        self.assertEqual(size_after, size_before)
+
+    def test_add_boot_PREP_full_resizes_larger(self):
```
Comment (dbungert) [inline subiquity/common/filesystem/tests/test_manipulator.py:415]: Obviously there are some similarities between the different bootloader tests.
What your thoughts on refactoring them a bit?

**Review by dbungert — COMMENTED** [review]
Maybe it's effective reverse psychology but this wasn't as bad as the description was making it sound.  I have several questions.

File: subiquity/common/filesystem/boot.py [hunk 2]
```diff
@@ -50,6 +55,227 @@ def _is_boot_device_raid(raid):
...
+
+@attr.s(auto_attribs=True)
+class ResizePlan(MakeBootDevicePlan):
+    """Resize a partition."""
+
+    part: object
+    size_delta: int = 0
+
+    def apply(self, manipulator):
+        assert not self.part.preserve
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/common/filesystem/boot.py:98]: I think this is more a check against the "can this be done" logic going wrong (perhaps this check should be done in a validator for the part field or something)

File: subiquity/common/filesystem/boot.py [hunk 3]
```diff
@@ -50,6 +55,227 @@ def _is_boot_device_raid(raid):
...
+
+    create_part_plan = CreatePartPlan(
+        device=device,
+        offset=None,
+        spec=dict(size=size, fstype='fat32', mount=None),
+        args=dict(flag='boot', grub_device=True))
+    if device._m._mount_for_path("/boot/efi") is None:
+        create_part_plan.spec['mount'] = '/boot/efi'
+
+    partitions = device.partitions()
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/common/filesystem/boot.py]: Yeah, can factor things a bit here.

File: subiquity/common/filesystem/tests/test_manipulator.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/common/filesystem/tests/test_manipulator.py:415]: Yeah they are pretty repetitive. I wonder if some well chosen utility functions might help... not today though.

File: subiquity/common/filesystem/boot.py [hunk 4]
```diff
@@ -50,6 +55,219 @@ def _is_boot_device_raid(raid):
...
+
+    if gaps.largest_gap_size(device) >= size:
+        create_part_plan.offset = gaps.largest_gap(device).offset
+        return create_part_plan
+    else:
+        largest_i, largest_part = max(
+            enumerate(partitions),
+            key=lambda i_p: i_p[1].size)
+        if size > largest_part.size // 2:
+            print(size, largest_part.size // 2)
```
Comment (dbungert) [inline subiquity/common/filesystem/boot.py:212]: Drop this line?

**Review by dbungert — APPROVED** [review]

File: subiquity/common/filesystem/boot.py (hunk 4 above)
Comment (mwhudson) [reply subiquity/common/filesystem/boot.py]: Done.

**Review by dbungert — APPROVED** [review]

**Pull request merged**